reader/reader_test.go

```diff
@@ -0,0 +1,138 @@
+package reader
+
+import (
+	"bytes"
+	"io"
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+)
+
+func TestRead(t *testing.T) {
+	data := []byte(`Header line
+First line
+Second line
+`)
+
+	Convey("Reader can read all data", t, func() {
+		inner := bytes.NewReader(data)
+		r := New(inner)
+
+		buf := make([]byte, 1024)
+		n, err := r.Read(buf)
+		So(err, ShouldEqual, nil)
+		So(n, ShouldEqual, len(data))
+		So(buf[:len(data)], ShouldResemble, data)
+		So(r.TotalBytesRead(), ShouldEqual, 35)
+	})
+
+	Convey("Reader can read all data after calling Peek", t, func() {
+		inner := bytes.NewReader(data)
+		r := New(inner)
+
+		header, err := r.PeekBytes('\n')
+		So(header, ShouldEqual, "Header line")
+		So(err, ShouldBeNil)
+
+		buf := make([]byte, 1024)
+		n, err := r.Read(buf)
+		So(err, ShouldEqual, nil)
+		So(n, ShouldEqual, len(data))
+		So(buf[:len(data)], ShouldResemble, data)
+	})
+
+	Convey("A second call to Read returns an error", t, func() {
+		inner := bytes.NewReader(data)
+		r := New(inner)
+
+		buf := make([]byte, 1024)
+		n, err := r.Read(buf)
+		So(err, ShouldEqual, nil)
+		So(n, ShouldEqual, len(data))
+		So(buf[:len(data)], ShouldResemble, data)
+
+		n, err = r.Read(buf)
+		So(err, ShouldEqual, io.EOF)
+		So(n, ShouldBeZeroValue)
+	})
+
+	Convey("Calling Read on empty reader returns an error", t, func() {
+		inner := bytes.NewReader(make([]byte, 0))
+		r := New(inner)
+
+		buf := make([]byte, 1024)
+		n, err := r.Read(buf)
+		So(err, ShouldEqual, io.EOF)
+		So(n, ShouldBeZeroValue)
+	})
+}
+
+func TestPeek(t *testing.T) {
+	data := []byte(`Header line
+First line
+Second line
+`)
+
+	upToS := []byte(`Header line
```

Comment:
Again something like this would look neater:
`upToS := []byte("Header line\nFirst line\n")`